docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,294 @@
+that the reverse proxy implementation to cause the lowest possible impact
+in the API server while processing high latency requests.
+
+Possible solutions:
+- reverse proxy implemented as an independent server
+- reverse proxy invoke backend services asynchronously
+
+----
+
+#### [G-2] ArgoCD admins should be able to define rbacs to define which users can invoke specific extensions
+
+ArgoCD Admins must be able to define which extensions are allowed to be
+executed by the logged in user. This should be fine grained by ArgoCD
+project like the current rbac implementation.
+
+----
+
+#### [G-3] ArgoCD deployment should be independent from backend services
+
+Extension developers should be able to deploy their backend services
+indepedendtly from ArgoCD. An extension can evolve their internal API and
+deploying a new version shouldn't require ArgoCD to be updated or
+restarted.
+
+----
+
+#### [G-4] Enhance the current Extensions framework to configure backend services
+
+*Not in the first release*
+
+[ArgoCD extensions][2] is an `argoproj-labs` project that supports loading
+extensions in runtime. Currently the project is implementing a controller
+that defines and reconciles the custom resource `ArgoCDExtension`. This
+CRD should be enhanced to provide the ability to define backend services
+that will be used by the extension. Once configured UI can send requests
+to API server in a specific endpoint. API server will act as a reverse
+proxy receiving the request from the UI and routing to the appropriate
+backend service.
+
+Example:
+```yaml 
+apiVersion: argoproj.io/v1alpha1
+kind: ArgoCDExtension
+metadata:
+  name: my-cool-extention
+  finalizers:
+    - extensions-finalizer.argocd.argoproj.io
+spec:
+  sources:
+    - git:
+        url: https://github.com/some-org/my-cool-extension.git
+  backend:
+    serviceName: some-backend-svc
+    endpoint: /some-backend
+```
+
+**Note**: While this is a nice-to-have, it won't be part of the first proxy
+extension version. This would need to be considered if ArgoCD extensions
+eventually get traction.
+
+## Non-Goals
+
+It isn't in the scope of this proposal to specify commands in the ArgoCD
+CLI. This proposal covers the reverse-proxy extension spec that will be
+used by ArgoCD UI.
+
+## Proposal
+
+### Use cases
+
+The following use cases should be implemented for the conclusion of this
+proposal:
+
+#### [UC-1]: As an ArgoCD admin, I want to configure a backend services so it can be used by my UI extension
+
+Define a new section in the ArgoCD configmap ([argocd-cm.yaml][4])
+allowing admins to register and configure new extensions. All enabled
+extensions backend will be available to be invoked by the ArgoCD UI under
+the following API base path:
+
+`<argocd-host>/api/v1/extensions/<extension-name>`
+
+
+With the configuration bellow, the expected behaviour is explained in the
+following examples:
+
+```yaml
+extension.config:
+| extensions:
```

Comment:
how the argocd would authenticate to the backend service?  Do we need to save the oauth token as a secret and pass the key in cm?

Reply:
For authentication ArgoCD will apply the preconfigured strategy in the API server ensuring that all incoming requests are coming from a valid subject.
For authorization it is defined in the UseCase 2 of this proposal how ArgoCD will leverage the current rbac to allow admins to configure access rights.